src/zkp/musig.rs

```diff
@@ -0,0 +1,1158 @@
+///! https://github.com/ElementsProject/scriptless-scripts/pull/24
+///!
+///! The documentation in this include file is for reference and may not be sufficient
+///! for users to begin using the library. A full description of the C API usage can be found
+///! in [C-musig.md](secp256k1-sys/depend/secp256k1/src/modules/musig/musig.md), and Rust API
+///! usage can be found in [Rust-musig.md](USAGE.md).
+use ffi::{self, CPtr};
+use schnorrsig::KeyPair;
+use Error;
+use Signing;
+use {Message, PublicKey, Secp256k1, SecretKey, Signature};
+
+///  Data structure containing auxiliary data generated in `pubkey_agg` and
+///  required for `session_*_init`.
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub struct MusigPreSession {
+    inner: ffi::MusigKeyaggCache,
+    agg_pk: ffi::XOnlyPublicKey,
+}
+
+impl CPtr for MusigPreSession {
+    type Target = ffi::MusigKeyaggCache;
+
+    fn as_c_ptr(&self) -> *const Self::Target {
+        self.as_ptr()
+    }
+
+    fn as_mut_c_ptr(&mut self) -> *mut Self::Target {
+        self.as_mut_ptr()
+    }
+}
+
+impl MusigPreSession {
+    /// Create a new MusigPreSession by supplying a list of PublicKeys used in the session
+    ///
+    /// Computes a combined public key and the hash of the given public keys.
+    ///
+    /// Different orders of `pubkeys` result in different `agg_pk`s.
+    ///
+    /// The pubkeys can be sorted before combining with `rustsecp256k1zkp_v0_4_0_xonly_sort` which
+    /// ensures the same resulting `agg_pk` for the same multiset of pubkeys.
+    /// This is useful to do before pubkey_combine, such that the order of pubkeys
+    /// does not affect the combined public key.
+    ///
+    /// Returns: MusigPreSession if the public keys were successfully combined, Error otherwise
+    /// Args:        secp: Secp256k1 context object initialized for verification
+    /// Out: pre_session: `MusigPreSession` struct to be used in
+    ///                   `MusigPreSession::nonce_process` or `MusigPreSession::pubkey_tweak_add`.
+    ///                   `MusigPreSession` also contains the Musig-combined xonly public key
+    ///  In:     pubkeys: input array of public keys to combine. The order
+    ///                   is important; a different order will result in a different
+    ///                   combined public key
+    ///
+    /// Example:
+    ///
+    /// ```rust
+    /// # use secp256k1_zkp::{MusigPreSession, PublicKey, Secp256k1, SecretKey};
+    /// let secp = Secp256k1::new();
+    /// let sec_bytes = [1; 32];
+    /// let sec_key = SecretKey::from_slice(&sec_bytes).unwrap();
+    /// let pub_key = PublicKey::from_secret_key(&secp, &sec_key);
+    /// let _pre_session = MusigPreSession::new(&secp, &[pub_key]).unwrap();
+    /// ```
+    pub fn new<C: Signing>(secp: &Secp256k1<C>, pubkeys: &[PublicKey]) -> Result<Self, Error> {
+        let cx = *secp.ctx();
+        let xonlys = pubkeys
+            .iter()
+            .map(|k| ffi::xonly_from_pubkey(cx, k.as_ptr()).0)
```

Comment:
Is there a particular reason for not directly using schnorr public keys which already are XOnly internally?